src/engine.rs

```diff
@@ -220,6 +231,8 @@ impl Engine {
     }
 
     fn consume_test_result(&mut self, result: TestResult) -> () {
+        assert!(self.state == EngineState::AwaitingCompletion);
+        self.state = EngineState::ReadyToProvide;
```

Comment:
Is it alright for this to change the state before doing the rest of the function?

Reply:
I think so. This can't exit early except through panicking. Maybe clearer to do it at the end though. Will change.

Reply:
Actually, there are a bunch of places where I use return. I think it's clearer if state is always updated at the beginning. The state machine here is basically that we need to have a sequence of paired calls `next_source` / `consume_test_result` and any other ordering is invalid, and the state is basically a single bit of information to verify that, so it's not used anywhere inside the function bodies.